### PR TITLE
Add a configuration to the cryptocurrency advanced service [ECR-4185]

### PR DESCRIPTION
**Pull request by aleksuss**
## Overview

---

See: https://jira.bf.local/browse/ECR-4185

### Definition of Done

- [ ] There are no TODOs left in the merged code
- [ ] Change is covered by automated tests
- [ ] Benchmark results are attached (if applicable)
- [ ] The [coding guidelines] are followed
- [ ] Public API has proper documentation
- [ ] Changelog is updated if needed (in case of notable or breaking changes)
- [ ] The continuous integration build passes

[coding guidelines]: https://github.com/exonum/exonum/blob/master/CONTRIBUTING.md#conventions

### CODE REVIEW

File: examples/cryptocurrency-advanced/backend/src/config.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+use super::proto;
+
+/// Cryptocurrency configuration parameters.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[derive(ProtobufConvert, BinaryValue)]
+#[protobuf_convert(source = "proto::Config")]
+pub struct Config {
+    /// Initial balance of the wallet.
+    pub init_balance: i64,
```
Comment (slowli) [inline examples/cryptocurrency-advanced/backend/src/config.rs]: Does using a signed integer only serve the purpose of getting an error during config verification? I'd suggest to use `u64` and check that `init_balance != 0`.

File: exonum/src/runtime/types.rs [hunk 2]
```diff
@@ -275,6 +275,16 @@ impl Display for ArtifactId {
     }
 }
 
+impl Debug for ArtifactId {
```
Comment (slowli) [inline exonum/src/runtime/types.rs]: Are this and the following `Debug` implementations motivated by the `non_exhaustive` marker? IMO, it's OK to output in `Debug`; it deals with the internal type structure anyway.

File: examples/cryptocurrency-advanced/backend/src/proto/service.proto [hunk 3]
```diff
@@ -57,4 +57,7 @@ message Wallet {
 }
 
 // Cryptocurrency configuration.
-message Config {}
+message Config {
+  // Initial balance of the wallet
```
Comment (slowli) [inline examples/cryptocurrency-advanced/backend/src/proto/service.proto]: Nit: I'd rephrase to "Initial balance of newly created wallets."

File: examples/cryptocurrency-advanced/backend/src/config.rs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            Ok(())
+        }
+    }
+}
+
+/// The enumeration represents errors that can occurred while validating config.
+#[derive(Debug, ExecutionFail)]
+pub enum ConfigError {
+    /// Initial balance is negative.
+    NegativeBalance = 1,
```
Comment (slowli) [inline examples/cryptocurrency-advanced/backend/src/config.rs]: The code of this error will clash with the `Error::SenderNotFound` variant defined in the `transactions` module. It could make sense to start enumeration here from 16 or 32 (and, ideally, explain in a comment why such separation is beneficial for forward compatibility).

File: examples/cryptocurrency-advanced/backend/tests/api.rs [hunk 5]
```diff
@@ -264,6 +268,55 @@ fn test_unknown_wallet_request() {
...
+
+    // Check that the user indeed is persisted by the service.
+    let wallet = api.get_wallet(tx.author()).unwrap();
+    assert_eq!(wallet.owner, author_address(&tx));
+    assert_eq!(wallet.name, ALICE_NAME);
+    assert_eq!(wallet.balance, 100);
+
+    // Change an initial balance to 200 via changing configuration.
+    let tx = api.change_config(&mut testkit, Config { init_balance: 200 });
+    testkit.create_block();
```
Comment (popzxc) [inline examples/cryptocurrency-advanced/backend/tests/api.rs]: Wouldn't it be simpler to use an existing API designed for this purpose:

```rust
let block = testkit.create_block();

assert!(block[tx.object_hash()].status().is_ok())
```

?

File: examples/cryptocurrency-advanced/backend/src/config.rs (hunk 4 above)
Comment (popzxc) [reply examples/cryptocurrency-advanced/backend/src/config.rs]: Note that such a separation is already introduced in [supervisor](https://github.com/exonum/exonum/blob/master/services/supervisor/src/errors.rs).

File: examples/cryptocurrency-advanced/backend/tests/api.rs (hunk 5 above)
Comment (slowli) [reply examples/cryptocurrency-advanced/backend/tests/api.rs]: To be fair, these are HTTP API tests, and `assert_tx_status` method is already used throughout the code.

File: examples/cryptocurrency-advanced/backend/src/lib.rs [hunk 6]
```diff
@@ -23,35 +23,62 @@ extern crate serde_derive; // Required for Protobuf.
...
     fn initialize(
         &self,
         context: ExecutionContext<'_>,
-        _params: Vec<u8>,
+        params: Vec<u8>,
     ) -> Result<(), ExecutionError> {
-        // Initialize indexes. Not doing this may lead to errors in HTTP API, since it relies on
-        // `wallets` indexes being initialized for returning corresponding proofs.
-        SchemaImpl::new(context.service_data());
+        let config = Config::from_bytes(params.into()).map_err(CommonError::malformed_arguments)?;
```
[the remark this answers is not in the document]
Comment (slowli) [reply examples/cryptocurrency-advanced/backend/src/lib.rs]: > AFAIK this `into` conversion will clone the vector data

Hm... Shouldn't the conversion consume `params` (i.e., simply wrap `Cow::Owned` variant around them)? Seems to be the case as per [this implementation](https://doc.rust-lang.org/src/alloc/vec.rs.html#2429-2433).

File: exonum/src/runtime/types.rs (hunk 2 above)
Comment (aleksuss) [reply exonum/src/runtime/types.rs]: But it appears in the log. This is a reason why i’ve reimplement Debug.

File: examples/cryptocurrency-advanced/backend/src/config.rs (hunk 1 above)
Comment (aleksuss) [reply examples/cryptocurrency-advanced/backend/src/config.rs]: Exactly

File: examples/cryptocurrency-advanced/backend/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (popzxc) [reply examples/cryptocurrency-advanced/backend/src/lib.rs]: Ah, indeed. Sorry, my bad.

File: examples/cryptocurrency-advanced/backend/tests/api.rs (hunk 5 above)
Comment (popzxc) [reply examples/cryptocurrency-advanced/backend/tests/api.rs]: Then I think the approach should be unified, since suggested approach is already used in this PR: https://github.com/exonum/exonum/pull/1760/files#diff-480f4bde234f6ab105cb6acf68052b6aR315

File: exonum/src/runtime/types.rs (hunk 2 above)
Comment (popzxc) [reply exonum/src/runtime/types.rs]: I think the problem here is that log entry uses debug representation (which is misuse). Log entries should use `Display`.

**Comment by slowli** [comment]
Another thing to note is that introducing config into the existing service will make the existing [advanced cryptocurrency tutorial](https://exonum.com/doc/version/latest/get-started/data-proofs/) clunky because it currently describes the initial wallet balance as a constant. I think the way around this would be place the version with config separately (I think it would be possible to reuse this newer version for the upcoming data migration tutorial). To minimize copy-paste, the reusable parts of the old service could be made public (with the comment explaining why) and would be used in the newer version.

File: examples/cryptocurrency-advanced/backend/src/schema.rs [hunk 7]
```diff
@@ -36,6 +36,8 @@ pub(crate) struct SchemaImpl<T: Access> {
     pub public: Schema<T>,
     /// History for specific wallets.
     pub wallet_history: Group<T, Address, ProofListIndex<T::Base, Hash>>,
+    /// Configuration of the service.
+    pub config: Entry<T::Base, Config>,
```
Comment (slowli) [inline examples/cryptocurrency-advanced/backend/src/schema.rs:40]: Is config present after a data migration? Seemingly not, which may lead to panics during wallet creation.

Comment (aleksuss) [reply examples/cryptocurrency-advanced/backend/src/schema.rs:40]: The configuration is being available only after migration to 0.2.0 version. Simple version of the cryptocurrency doesn't have a configuration.
